doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+a class is being loaded, the JIT runs through all the relevant runtime assumptions and patches the locations.
+
+Virtual guard noping can be disabled by setting `-Xjit:disableVirtualGuardNOPing`. There are also cases where
+we do not want to use nopable guard. During start-up, all kinds of classes are still being loaded and the class
+hierarchy has not settled down. By using nopable guard too early on, we would pay the price for patching because
+after the guard is patched, the execution goes to the slow path forever. In this case, we might choose to be more
+conservative and use explicit guards although they are less efficient. At least some proportion of the time,
+the execution could go to the inlined path.
+
+## Single Abstract Implementation Guard (Nopable guard)
+
+When there is only one concrete class extending the abstract class and there is a call on an abstract class,
+there is no need for profiling to tell which object or which method is going to be executed. It is going to be
+ that single concrete implementation since there is no other concrete implementation. It gets patched when
+there is a new concrete implementation of that abstract class is loaded. Meanwhile it is also difficult to do
+a test because we do not want to walk through all the subclasses of the abstract class to figure out how many
+entries there are and which entry it should be.  It could also get expensive to do the test if the method just
+returns a constant field.
+
+## Single Interface Implementation Guard (Nopable Guard)
+
+Similarly, single interface implementation guard is added when there is a call on an interface and there is only
+one implementation of the interface. The single implementer is inlined. In the future, when a new implementation
+gets loaded that violates the runtime assumption, the single interface implementation guard is patched.
+
+## Hierarchy Guard (Nopable Guard)
+
+For example, there is a call `Object.equals()` in the bytecode. There are tons of classes that could implement
+`equals()`. If JIT knows the call is invoked on an object of `Map` or an object of a subclass of `Map` through
+value propagation and type propagation, a nopable guard can be used as long as there is only one implementation
+of `Map` at compile time. If a new implementation gets loaded, the code is patched, and the slow path will be
+invoked. In this case, although `Map.equals()` is not in the bytecode, the nopable guard can be generated based
+on the information learned through value propagation and type propagation.
+
+## Inner Preexistence Guard
+
+Inner preexistence guard is a single guard that combines multiple guards that exist for other calls within the
+method. When a method is inlined, a guard is added around it. Within that inlined code, there could be a need
+to apply more guards such as inlining other code. The inner preexistence guard pushes some or all the guards
```

Comment:
The key aspect of inner preexistence is that it only merges assumptions/conditions from inner calls that had a callee parameter as the receiver used in the call, i.e. they would have been devirtualized using normal preexistence if they had not been inlined.